### PR TITLE
Qwen3

### PR DESCRIPTION
**Pull request by cuichenx**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by athitten** [comment]
Approving the changes in `nemo/collections/llm/api.py`. Thanks @cuichenx!

**Review by pablo-garay — APPROVED** [review]

File: nemo/collections/llm/gpt/model/qwen3.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+class Qwen3MoEConfig(Qwen3Config):
+    """
+    Base config for Qwen 3 MoE Models
+    """
+
+    num_moe_experts: int = 128
+    moe_router_load_balancing_type: str = "aux_loss"
+    moe_aux_loss_coeff: float = 1e-3
+    moe_router_topk: int = 8
+    moe_router_pre_softmax: bool = False
```
Comment (skyshine102) [inline nemo/collections/llm/gpt/model/qwen3.py:62]: Hi, I have a question regarding `moe_router_pre_softmax` argument. 

The argument is supposed to be set to `True` since Qwen3 uses F.softmax after Linear MoE gate.

Comment (cuichenx) [reply nemo/collections/llm/gpt/model/qwen3.py:62]: Hi @skyshine102, thanks for your question! This is a good question
Qwen3 sets `norm_topk_prob` to True, and this applies an normalization after topk. Doing softmax -- topk -- norm is mathematically equivalent to topk -- softmax. Hence we set `moe_router_pre_softmax` to False

Comment (skyshine102) [reply nemo/collections/llm/gpt/model/qwen3.py:62]: Hi @cuichenx, I'm still confused. 
I'm not that familiar with Mcore implementation. But from the model configs of [Mixtral](https://github.com/NVIDIA/NeMo/blob/822247934381af0f4d84e3aacac7a2558d1b2e35/nemo/collections/llm/gpt/model/mixtral.py#L62C5-L62C27) and [Deepseek v3](https://github.com/NVIDIA/NeMo/blob/822247934381af0f4d84e3aacac7a2558d1b2e35/nemo/collections/llm/gpt/model/deepseek.py#L89C5-L89C40), all of them set `moe_router_pre_softmax` to True.

Their corresponding implementation both have softmax -- topk -- norm. 
See
- HF Mixtral (though no `norm_topk_prob` appears in hf config): 
https://github.com/huggingface/transformers/blob/7b7bb8df9793f983ec3aaf7a43a3387d587f2182/src/transformers/models/mixtral/modeling_mixtral.py#L120
- HF Deepseek v3 (have`norm_topk_prob` in hf config)
https://huggingface.co/deepseek-ai/DeepSeek-V3/blob/main/modeling_deepseek.py#L469

Can you help me out?

Comment (cuichenx) [reply nemo/collections/llm/gpt/model/qwen3.py:62]: In the case of DeepSeek, V3 uses sigmoid scoring function, so that argument has no effect. The sigmoid implementation in MCore has the normalization operation ([see here](https://github.com/nvidia/Megatron-LM/blob/main/megatron/core/transformer/moe/moe_utils.py#L557-L565)). The NeMo config is there for DeepSeek V2 which has norm_topk_prob=False. 

For Mixtral, I'm confirming with the author of that model. I will get back to you.

Comment (cuichenx) [reply nemo/collections/llm/gpt/model/qwen3.py:62]: I think you found a bug in our mixtral implementation

Comment (cuichenx) [reply nemo/collections/llm/gpt/model/qwen3.py:62]: Fix is in this PR: https://github.com/NVIDIA/NeMo/pull/13678

Comment (skyshine102) [reply nemo/collections/llm/gpt/model/qwen3.py:62]: Thanks for the detailed explanation! Glad to catch a bug lol.

> In the case of DeepSeek, V3 uses sigmoid scoring function, so that argument has no effect in MCore. The sigmoid implementation in MCore has the normalization operation ([see here](https://github.com/nvidia/Megatron-LM/blob/main/megatron/core/transformer/moe/moe_utils.py#L557-L565)). The NeMo config has `moe_router_pre_softmax` set to True because of DeepSeek V2, which has norm_topk_prob=False.
> 
> For Mixtral, I'm confirming with the author of that model. I will get back to you.

I would like to discuss a bit more on the models using sigmoid scoring function:
- [Llama 4](https://github.com/NVIDIA/NeMo/blob/c882b4885349b5a750147d242338ebcbc7058ef6/nemo/collections/llm/gpt/model/llama.py#L399) set `moe_router_pre_softmax: bool = False` (no effect), `moe_router_score_function = "sigmoid"`
    - Llama 4 HF/Meta repo implements: Linear --> topk --> fill "-inf" --> sigmoid
    - This seems to be not the same as Linear --> sigmoid --> topk --> norm to 1 (which moe_router_score_function= "sigmoid" means)
 
 Please correct me if I'm wrong.

File: nemo/collections/llm/gpt/model/qwen3.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        output = qwen3_config_cls(
+            num_layers=source.num_hidden_layers,
+            hidden_size=source.hidden_size,
+            ffn_hidden_size=source.intermediate_size,
+            num_attention_heads=source.num_attention_heads,
+            num_query_groups=source.num_key_value_heads,
+            init_method_std=source.initializer_range,
+            layernorm_epsilon=source.rms_norm_eps,
+            vocab_size=source.vocab_size,
+            make_vocab_size_divisible_by=1187,
```
Comment (leoleoasd) [inline nemo/collections/llm/gpt/model/qwen3.py:293]: Hi @cuichenx, I see in other models this field is 128 or other power of 2. Is there a specific reason to use `1187` here?
